support-workers/src/test/scala/com/gu/support/workers/integration/CreateZuoraSubscriptionSpec.scala

```diff
@@ -50,51 +52,59 @@ class CreateZuoraSubscriptionSpec extends AsyncLambdaSpec with MockServicesCreat
     for {
       _ <- setCodeStatus(mutableCode, RedemptionTable.AvailableField.CodeIsAvailable)
       _ <- createSubscription(createDigiPackCorporateSubscriptionJson)
-      r <- getCodeStatus(mutableCode).map { _ should be(Left(GetCodeStatus.CodeAlreadyUsed))}
+      r <- getCodeStatus(mutableCode).map {
+        _ should be(Left(GetCodeStatus.CodeAlreadyUsed))
+      }
     } yield r
   }
 
   it should "create a Digital Pack subscription with a discount" in {
-    createSubscription(createDigiPackSubscriptionWithPromoJson)
+    testCreateSubscription(createDigiPackSubscriptionWithPromoJson)
   }
 
   it should "create a Digital Pack subscription with a discount and free trial" in {
-    createSubscription(digipackSubscriptionWithDiscountAndFreeTrialJson)
-  }
-
-  it should "create a Digital Pack gift subscription" in {
-    createSubscription(createDigiPackGiftSubscriptionJson)
+    testCreateSubscription(digipackSubscriptionWithDiscountAndFreeTrialJson)
   }
 
   it should "create an everyday paper subscription" in {
-    createSubscription(createEverydayPaperSubscriptionJson)
+    testCreateSubscription(createEverydayPaperSubscriptionJson)
   }
 
   it should "create an Annual Guardian Weekly subscription" in {
-    createSubscription(createGuardianWeeklySubscriptionJson(Annual))
+    testCreateSubscription(createGuardianWeeklySubscriptionJson(Annual))
   }
 
   it should "create an Quarterly Guardian Weekly subscription" in {
-    createSubscription(createGuardianWeeklySubscriptionJson(Quarterly))
+    testCreateSubscription(createGuardianWeeklySubscriptionJson(Quarterly))
   }
 
   it should "create a 6 for 6 Guardian Weekly subscription" in {
-    createSubscription(createGuardianWeeklySubscriptionJson(SixWeekly, Some(DefaultPromotions.GuardianWeekly.NonGift.sixForSix)))
+    testCreateSubscription(createGuardianWeeklySubscriptionJson(SixWeekly, Some(DefaultPromotions.GuardianWeekly.NonGift.sixForSix)))
   }
 
   it should "create an Guardian Weekly gift subscription" in {
-    createSubscription(guardianWeeklyGiftJson)
+    testCreateSubscription(guardianWeeklyGiftJson)
   }
 
-  private def createSubscription(json: String) = {
-    val createZuora = new CreateZuoraSubscription(mockServiceProvider)
+  def testCreateSubscription(json: String) =
+    createSubscription(json)
+      .map(
+        _._1.subscriptionNumber.length should be > 1
+      )
+}
+
+trait CreateZuoraSubscriptionBaseSpec extends AsyncLambdaSpec with MockServicesCreator with MockContext {
+
+  def createSubscription(
```

Comment:
I agree in modularising this better, but why put these in a trait rather than an object?  if in an object then it is self contained/can't depend on anything in the subclass unless it's passed in.  trait doesn't have that guarantee.

Reply:
I had them in an Object originally but it caused at least one test to hang indefinitely - I think it was to do with the way that Scalatest uses its own execution context for async tests. Converting to a trait solved this.

Reply:
if it really has some state that breaks things then consider make it its own class rather than extending, that's still simpler/self contained than extending.

We can revisit, but we did have in the [coding standards](https://docs.google.com/document/d/1vRIYAU2OnRkHzf4dLCuzxiA-3HsKh1PDmkiLc-1GmfQ/edit#) to avoid inheritance and overriding, in favour of normal functions.  But I think if the hanging problem is impossible, creating a standalone class would be a better middle ground as there's not much unexpected a composed class can do.

Reply:
👍